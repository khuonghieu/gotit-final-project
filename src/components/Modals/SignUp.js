import React, { useState } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import {
  Button, Form, Modal, Message,
} from '@gotitinc/design-system';
import { signUp } from '../../actions/users';

const mapDispatchToProps = (dispatch) => ({
  signUp: (email, username, password, name) => dispatch(signUp(email, username, password, name)),
});

export function SignUpModal({ onClose, signUp }) {
  const [email, setEmail] = useState('');
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const [name, setName] = useState('');

  const [errorMessage, setErrorMessage] = useState();

  async function handleSignup(e) {
    e.preventDefault();
    if (email && username && password && name) {
      const { success, payload } = await signUp(email, username, password, name);
      if (!success) {
        setErrorMessage(JSON.stringify(payload.message));
      }
    } else {
      alert('fill all the blanks');
    }
  }

  return (
    <div>
      <Modal show centered size="small" onHide={onClose}>
        <Modal.Header closeButton onClick={onClose}>
          <Modal.Title>Sign Up</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          {errorMessage ? (
            <Message type="system" variant="negative">
              <Message.Container>
                <Message.Title>
                  Sign up failed
                </Message.Title>
                <Message.Content>
                  {errorMessage}
                </Message.Content>
              </Message.Container>
            </Message>
          ) : null}
          <div className="u-textCenter">
            <img src="holder.js/100px90?text=Image" className="u-maxWidthFull u-marginBottomExtraSmall" alt="" />
          </div>
          <Form.Label>Email</Form.Label>
          <Form.Input value={email} onChange={(e) => setEmail(e.target.value)} type="email" placeholder="Enter email" name="email" />
          <Form.Label>Username</Form.Label>
          <Form.Input value={username} onChange={(e) => setUsername(e.target.value)} type="text" placeholder="Enter username" name="username" />
          <Form.Label>Password</Form.Label>
          <Form.Input value={password} onChange={(e) => setPassword(e.target.value)} type="password" placeholder="Enter password" name="password" />
          <Form.Label>Name</Form.Label>
          <Form.Input value={name} onChange={(e) => setName(e.target.value)} type="text" placeholder="Enter your name" name="name" />
          <Button onClick={handleSignup}>Sign Up</Button>
        </Modal.Body>
      </Modal>
    </div>
  );
}

SignUpModal.propTypes = {
  onClose: PropTypes.func,
  signUp: PropTypes.func,
};

export default connect(null, mapDispatchToProps)(SignUpModal);
