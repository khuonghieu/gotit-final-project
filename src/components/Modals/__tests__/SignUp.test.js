import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Form, Button, Message } from '@gotitinc/design-system';
import { SignUpModal } from '../SignUp';

configure({ adapter: new Adapter() });

describe('components/Modals/SignUp.js', () => {
  let wrapper;
  let props;
  let input;
  let button;

  const update = () => {
    input = wrapper.find(Form.Input);
    button = wrapper.find(Button);
  };

  const setup = () => {
    wrapper = shallow(
      <SignUpModal {...props} />,
    );
    update();
  };
  beforeEach(() => {
    props = {
      onClose: jest.fn(),
      signUp: jest.fn(),
    };
  });

  it('should render correctly', () => {
    setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should call SignUn API when form is fully filled and clicked submit button', () => {
    setup();
    input.at(0).simulate('change', { target: { value: 'testemail@gmail.com' } });
    input.at(1).simulate('change', { target: { value: 'testusername' } });
    input.at(2).simulate('change', { target: { value: 'testpassword' } });
    input.at(3).simulate('change', { target: { value: 'testname' } });
    update();
    button.props().onClick({ preventDefault: jest.fn() });
    expect(props.signUp).toHaveBeenCalled();
  });

  it('should display error message when all fields are not fully filled', () => {
    setup();
    button.props().onClick({ preventDefault: jest.fn() });
    expect(wrapper.find(Message).length).toBe(1);
  });
  it('should display error message when sign up failed', async () => {
    props.signUp.mockReturnValue({ success: false, payload: { message: 'error' } });
    setup();
    input.at(0).simulate('change', { target: { value: 'testemail@gmail.com' } });
    input.at(1).simulate('change', { target: { value: 'testusername' } });
    input.at(2).simulate('change', { target: { value: 'testpassword' } });
    input.at(3).simulate('change', { target: { value: 'testname' } });
    update();
    await button.props().onClick({ preventDefault: jest.fn() });
    expect(wrapper.find(Message).length).toBe(1);
  });
});
