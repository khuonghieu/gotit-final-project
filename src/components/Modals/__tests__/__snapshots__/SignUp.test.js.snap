// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Modals/SignUp.js should render correctly 1`] = `ShallowWrapper {}`;
